numba/parfors/array_analysis.py

```diff
@@ -89,22 +89,34 @@ def wrap_index(typingctx, idx, size):
     (idx < 0 ? idx + size : idx) because we may have situations
     where idx > size due to the way indices are calculated
     during slice/range analysis.
+
+    Both idx and size have to be Integer types.
+    size should be from the array size vars that array_analysis
+    adds and the bitwidth should match the platform maximum.
     """
-    unified_ty = typingctx.unify_types(idx, size)
-    # Mixing signed and unsigned ints will unify to double which is
-    # no good for indexing.  If the unified type is not an integer
-    # then just use int64 as the common index type.  This does have
-    # some overflow potential if the unsigned value is greater than
-    # 2**63.
-    if not isinstance(unified_ty, types.Integer):
-        unified_ty = types.int64
+    require(isinstance(idx, types.scalars.Integer))
+    require(isinstance(size, types.scalars.Integer))
+
+    # We need both idx and size to be platform size so that we can compare.
+    unified_ty = types.intp if size.signed else types.uintp
+    idx_unified = types.intp if idx.signed else types.uintp
 
     def codegen(context, builder, sig, args):
+        ll_idx_unified_ty = context.get_data_type(idx_unified)
         ll_unified_ty = context.get_data_type(unified_ty)
-        idx = builder.sext(args[0], ll_unified_ty)
-        size = builder.sext(args[1], ll_unified_ty)
+        if idx_unified.signed:
+            idx = builder.sext(args[0], ll_idx_unified_ty)
+        else:
+            idx = builder.zext(args[0], ll_idx_unified_ty)
+        if unified_ty.signed:
+            size = builder.sext(args[1], ll_unified_ty)
+        else:
+            size = builder.zext(args[1], ll_unified_ty)
```

Comment:
This branch is not hit in testing. Can it be hit?

Reply:
One would think it would be more likely for a size to be unsigned that it is to be signed so if 112 is hit instead of 114 then that is a bit surprising.  I would think that both could happen but I'd have to trace all the callsites in the code to see where the parameters passed as size are coming from.  I don't know off the top of my head how to write a test to force this to happen.

Reply:
hmmm, the test suite suggests the converse, size is always signed (and `int64`)? From inspection this branch is correct, so I'm happy to leave it, but I do wonder if it's actually possible to hit it. Any objections to merging as-is?

Reply:
Yes, let's just leave it.  Thanks.